pep-9999.rst

```diff
@@ -0,0 +1,247 @@
+        ...
+
+    # Or equivalently
+    m : typing.Match
+    if m := pattern.match(data):
+        ...
+    elif m := other_pattern.match(data):
+        ...
+    elif m := yet_naother_pattern.match(data):
+        ...
+    else:
+        ...
+
+
+Note that this PEP also delegates the task of deciding whether or not it's a
+good idea to implicitly allow nonlocal name rebinding operations to PEP 572 -
+it's a competing PEP aimed at ensuring the resulting capability is at least
+moderately teachable (think "If we do this at all, we should do it this
+particular way, and explain it to end users with this particular semantic
+framing" rather than seeing it as an independent proposal), and the PEP author
+would be entirely happy with an outcome where both PEPs ended up being rejected
+(as if we decide "not right now", it's straightforward to reconsider inline
+assignment support at a later date, while once we accept and release it, we're
+going to be stuck with it indefinitely).
+
+
+Syntax and semantics
+====================
+
+Augmented assignment expressions
+--------------------------------
+
+The language grammar would be adjusted to allow augmented assignments to appear
+as expressions, where the result of the expression is the same reference that
+is being bound to the given target.
+
+When used as anything other than a standalone expression statement, the
+assignment target would be restricted to a simple name (see design discussion
+for more background on this).
+
+
+Augmented assignment target scoping
+-----------------------------------
+
+If augmented assignment expressions were added with their current scoping
+semantics (where they implicitly bind a local name, just like a regular
+assignment statement), then code like ``(total += value for value in data)``
+would fail at runtime with ``UnboundLocalError``, just as the following code
+does today::
+
+    def cumulative_sums(data, start=0)
+        total = start
+        def _inner_gen():
+            for value in data:
+                total += value
+                yield total
+        yield from _inner_gen()
+        return total
+
+To address this, the compilation process will be changed to handle augmented
+assignment target scope resolution as follows:
+
+1. If the augmented assignment appears in a scope that does not participate in
+   lexically scoped name resolution (e.g. module globals or a class body), then
+   the target is always bound as a local variable
+2. If the augmented assignment target is covered in the current lexical scope
+   by an explicit scope declaration (either ``global`` or ``nonlocal``) then
+   that scope is used.
+3. If the augmented assignment target is explicitly declared as local by other
+   means (i.e. a function parameter declaration, a variable declaration, a
+   regular assignment statement, or another name binding operation), then the
+   local scope is used.
+4. If implicitly declaring the augmented assignment target as ``nonlocal`` would
```

Comment:
Yep, and if it bothers Guido too much, then we can either restrict doing this implicitly to scopes that are themselves implicit (i.e. comprehensions and generator expressions), or else restrict it to one level of implicit scoping, and require explicit scoping beyond that (which would still be enough to allow `add_one = lambda: tally += 1`).

Either of those two variants have the virtue of playing nice with global scopes, as well as preventing implicitly "jumping over" class scopes (you'd have to explicitly jump over them by declaring a name as explicitly non-local in a method body).

Reply:
Any extra complexity in the scoping rules beyond what you've proposed is going to make this a harder sell.  So I'd be careful there.  It might be good enough to add something like a ``__future__.implicit_nonlocal`` for just this case, to bring an extra measure of awareness and to give folks a chance to check for (unlikely) backward incompatibility.